File: apps/host/internal/githubhost/references.go
package githubhost

import (
	"context"
	"strings"

	pb "armadra.local/host/gen/armadra/v1"
	"armadra.local/host/internal/storage"
)

// External references link one Issue or pull request to one local session,
// branch or worktree (design §7.1). A reference is a badge and a way back — it
// never turns a session into a GitHub object, and unlinking never touches
// either side.

func referenceKind(value int64) pb.GithubReferenceKind {
	if value == storage.GithubReferencePull {
		return pb.GithubReferenceKind_GITHUB_REFERENCE_KIND_PULL_REQUEST
	}
	return pb.GithubReferenceKind_GITHUB_REFERENCE_KIND_ISSUE
}

func referenceKindValue(kind pb.GithubReferenceKind) (int64, error) {
	switch kind {
	case pb.GithubReferenceKind_GITHUB_REFERENCE_KIND_ISSUE:
		return storage.GithubReferenceIssue, nil
	case pb.GithubReferenceKind_GITHUB_REFERENCE_KIND_PULL_REQUEST:
		return storage.GithubReferencePull, nil
	}
	return 0, ErrInvalid
}

func targetKind(value int64) pb.GithubReferenceTargetKind {
	switch value {
	case storage.GithubTargetBranch:
		return pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_BRANCH
	case storage.GithubTargetWorktree:
		return pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_WORKTREE
	}
	return pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_SESSION
}

func targetKindValue(kind pb.GithubReferenceTargetKind) (int64, error) {
	switch kind {
	case pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_SESSION:
		return storage.GithubTargetSession, nil
	case pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_BRANCH:
		return storage.GithubTargetBranch, nil
	case pb.GithubReferenceTargetKind_GITHUB_REFERENCE_TARGET_KIND_WORKTREE:
		return storage.GithubTargetWorktree, nil
	}
	return 0, ErrInvalid
}

func referenceMessage(record storage.GithubReferenceRecord) *pb.GithubExternalReference {
	return &pb.GithubExternalReference{
		ReferenceId: record.ReferenceID,
		WorkspaceId: record.WorkspaceID,
		Repository: &pb.GithubRepositoryRef{
			Owner:   record.Repository.Owner,
			Name:    record.Repository.Name,
			ApiBase: record.Repository.APIBase,
			Host:    record.Repository.WebHost,
		},
		Kind:            referenceKind(record.Kind),
		Number:          record.Number,
		TargetKind:      targetKind(record.TargetKind),
		TargetId:        record.TargetID,
		Title:           record.Title,
		Revision:        record.Revision,
		CreatedAtUnixMs: record.CreatedAtMS,
		UpdatedAtUnixMs: record.UpdatedAtMS,
	}
}

// referencesFor collects the links pointing at one remote object, so a detail
// view can show which local sessions or worktrees are working on it.
func (s *Service) referencesFor(ctx context.Context, workspace string, ref *pb.GithubRepositoryRef, kind pb.GithubReferenceKind, number int64) ([]*pb.GithubExternalReference, error) {
	records, err := s.store.GithubReferences(ctx, workspace, "", "", MaxPageLimit)
	if err != nil {
		return nil, err
	}
	result := []*pb.GithubExternalReference{}
	for _, record := range records {
		if record.Number != number || referenceKind(record.Kind) != kind {
			continue
		}
		if record.Repository.Owner != ref.Owner || record.Repository.Name != ref.Name || record.Repository.APIBase != ref.ApiBase {
			continue
		}
		result = append(result, referenceMessage(record))
	}
	return result, nil
}

// LinkReference records one link. Its identifier is derived from what the link
// means, so linking the same Issue to the same target twice is the same record
// rather than two badges on one node.
func (s *Service) LinkReference(ctx context.Context, caller Caller, request *pb.LinkGithubReferenceRequest) (*pb.GithubExternalReference, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	reference := request.GetReference()
	if reference == nil {
		return nil, ErrInvalid
	}
	ref, err := s.repository(ctx, reference.GetRepository())
	if err != nil {
		return nil, err
	}
	kind, err := referenceKindValue(reference.GetKind())
	if err != nil {
		return nil, err
	}
	target, err := targetKindValue(reference.GetTargetKind())
	if err != nil {
		return nil, err
	}
	id := strings.TrimSpace(reference.GetTargetId())
	if id == "" || len(id) > 512 || reference.GetNumber() <= 0 || len(reference.GetTitle()) > 1024 {
		return nil, ErrInvalid
	}
	now := s.now().UnixMilli()
	record := storage.GithubReferenceRecord{
		ReferenceID: referenceID(caller.WorkspaceID, ref, reference.GetKind(), reference.GetNumber(), reference.GetTargetKind(), id),
		WorkspaceID: caller.WorkspaceID,
		Repository:  key(ref),
		Kind:        kind,
		Number:      reference.GetNumber(),
		TargetKind:  target,
		TargetID:    id,
		Title:       reference.GetTitle(),
		CreatedAtMS: now,
		UpdatedAtMS: now,
	}
	stored, err := s.store.PutGithubReference(ctx, record, request.GetExpectedRevision())
	if err != nil {
		return nil, err
	}
	return referenceMessage(stored), nil
}

// UnlinkReference removes one link only. The remote object and the local
// session it pointed at are both left exactly as they were.
func (s *Service) UnlinkReference(ctx context.Context, caller Caller, request *pb.UnlinkGithubReferenceRequest) (*pb.UnlinkGithubReferenceResponse, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	id := request.GetReferenceId()
	if id == "" || len(id) > 256 || request.GetExpectedRevision() == 0 {
		return nil, ErrInvalid
	}
	if err := s.store.DeleteGithubReference(ctx, caller.WorkspaceID, id, request.GetExpectedRevision()); err != nil {
		return nil, err
	}
	return &pb.UnlinkGithubReferenceResponse{ReferenceId: id, Unlinked: true}, nil
}

// ListReferences lists this workspace's links, optionally for one target. An
// empty target lists the whole workspace; it is never an implicit match.
func (s *Service) ListReferences(ctx context.Context, caller Caller, request *pb.ListGithubReferencesRequest) (*pb.ListGithubReferencesResponse, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	limit := pageLimit(request.GetLimit())
	records, err := s.store.GithubReferences(ctx, caller.WorkspaceID, request.GetTargetId(), request.GetAfterId(), limit+1)
	if err != nil {
		return nil, err
	}
	result := &pb.ListGithubReferencesResponse{}
	for index, record := range records {
		if index == limit {
			result.HasMore = true
			break
		}
		result.References = append(result.References, referenceMessage(record))
		result.NextId = record.ReferenceID
	}
	return result, nil
}

// Credential wrappers. The status they return never contains a token; only a
// configure request carries one, inbound.

func (s *Service) GetCredential(ctx context.Context, caller Caller) (*pb.GithubCredentialStatus, error) {
	if err := s.authorize(caller, ScopeRead); err != nil {
		return nil, err
	}
	return s.credentials.Status(ctx)
}

// ConfigureCredential requires github:write and settings authority together:
// choosing where this Host's GitHub token comes from is a settings change, not
// an ordinary Issue edit.
func (s *Service) ConfigureCredential(ctx context.Context, caller Caller, request *pb.ConfigureGithubCredentialRequest) (*pb.GithubCredentialStatus, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	if err := s.authorize(caller, "settings:write"); err != nil {
		return nil, err
	}
	return s.credentials.Configure(ctx, request.GetSource(), request.GetToken(), request.GetApiBase(), request.GetExpectedRevision())
}

func (s *Service) RevokeCredential(ctx context.Context, caller Caller, request *pb.RevokeGithubCredentialRequest) (*pb.GithubCredentialStatus, error) {
	if err := s.authorize(caller, ScopeWrite); err != nil {
		return nil, err
	}
	if err := s.authorize(caller, "settings:write"); err != nil {
		return nil, err
	}
	return s.credentials.Revoke(ctx, request.GetExpectedRevision())
}

File: apps/desktop/src/core/github/references.ts
/**
 * 外部连接：把一条 Issue 或 PR 连到一个本地会话、分支或 worktree（设计 §7.1）。
 * 移植自 `apps/host/internal/githubhost/references.go`。
 *
 * 一条连接是一个徽标和一条回去的路——它从不把一个会话变成一个 GitHub 对象，取消
 * 连接也从不碰两边任何一边。
 */

import {
  GithubExternalReferenceSchema,
  GithubReferenceKind,
  GithubReferenceTargetKind,
  GithubRepositoryRefSchema,
  ListGithubReferencesResponseSchema,
  UnlinkGithubReferenceResponseSchema,
  create,
  type GithubCredentialStatus,
  type GithubExternalReference,
  type GithubRepositoryRef,
  type ConfigureGithubCredentialRequest,
  type LinkGithubReferenceRequest,
  type ListGithubReferencesRequest,
  type ListGithubReferencesResponse,
  type RevokeGithubCredentialRequest,
  type UnlinkGithubReferenceRequest,
  type UnlinkGithubReferenceResponse,
} from "@armadra/protocol";

import { githubError } from "./errors";
import {
  MAX_PAGE_LIMIT,
  SCOPE_READ,
  SCOPE_WRITE,
  key,
  pageLimit,
  referenceId,
  type Caller,
  type GithubService,
} from "./service";
import {
  GITHUB_REFERENCE_ISSUE,
  GITHUB_REFERENCE_PULL,
  GITHUB_TARGET_BRANCH,
  GITHUB_TARGET_SESSION,
  GITHUB_TARGET_WORKTREE,
  type GithubReferenceRecord,
} from "./store";

function referenceKind(value: number): GithubReferenceKind {
  return value === GITHUB_REFERENCE_PULL
    ? GithubReferenceKind.PULL_REQUEST
    : GithubReferenceKind.ISSUE;
}

function referenceKindValue(kind: GithubReferenceKind): number {
  if (kind === GithubReferenceKind.ISSUE) return GITHUB_REFERENCE_ISSUE;
  if (kind === GithubReferenceKind.PULL_REQUEST) return GITHUB_REFERENCE_PULL;
  throw githubError("invalid");
}

function targetKind(value: number): GithubReferenceTargetKind {
  if (value === GITHUB_TARGET_BRANCH) return GithubReferenceTargetKind.BRANCH;
  if (value === GITHUB_TARGET_WORKTREE) {
    return GithubReferenceTargetKind.WORKTREE;
  }
  return GithubReferenceTargetKind.SESSION;
}

function targetKindValue(kind: GithubReferenceTargetKind): number {
  switch (kind) {
    case GithubReferenceTargetKind.SESSION:
      return GITHUB_TARGET_SESSION;
    case GithubReferenceTargetKind.BRANCH:
      return GITHUB_TARGET_BRANCH;
    case GithubReferenceTargetKind.WORKTREE:
      return GITHUB_TARGET_WORKTREE;
    default:
      throw githubError("invalid");
  }
}

export function referenceMessage(
  record: GithubReferenceRecord,
): GithubExternalReference {
  return create(GithubExternalReferenceSchema, {
    referenceId: record.referenceId,
    workspaceId: record.workspaceId,
    repository: create(GithubRepositoryRefSchema, {
      owner: record.repository.owner,
      name: record.repository.name,
      apiBase: record.repository.apiBase,
      host: record.repository.webHost,
    }),
    kind: referenceKind(record.kind),
    number: BigInt(record.number),
    targetKind: targetKind(record.targetKind),
    targetId: record.targetId,
    title: record.title,
    revision: BigInt(record.revision),
    createdAtUnixMs: BigInt(record.createdAtMs),
    updatedAtUnixMs: BigInt(record.updatedAtMs),
  });
}

/**
 * 收集指向一个远端对象的连接，这样详情视图可以显示哪些本地会话或 worktree 正在
 * 做它。
 */
export function referencesFor(
  service: GithubService,
  workspaceId: string,
  ref: GithubRepositoryRef,
  kind: GithubReferenceKind,
  number: bigint,
): GithubExternalReference[] {
  const records = service.store.references(workspaceId, "", "", MAX_PAGE_LIMIT);
  const result: GithubExternalReference[] = [];
  for (const record of records) {
    if (
      BigInt(record.number) !== number ||
      referenceKind(record.kind) !== kind
    ) {
      continue;
    }
    if (
      record.repository.owner !== ref.owner ||
      record.repository.name !== ref.name ||
      record.repository.apiBase !== ref.apiBase
    ) {
      continue;
    }
    result.push(referenceMessage(record));
  }
  return result;
}

/**
 * 记下一条连接。它的标识由这条连接的含义导出，所以把同一个 Issue 连到同一个目标
 * 两次是同一条记录，而不是一个节点上的两个徽标。
 */
export function linkReference(
  service: GithubService,
  caller: Caller,
  request: LinkGithubReferenceRequest,
): GithubExternalReference {
  service.authorize(caller, SCOPE_WRITE);
  const reference = request.reference;
  if (reference === undefined) throw githubError("invalid");
  const ref = service.repository(reference.repository);
  const kind = referenceKindValue(reference.kind);
  const target = targetKindValue(reference.targetKind);
  const id = reference.targetId.trim();
  if (
    id === "" ||
    id.length > 512 ||
    reference.number <= 0n ||
    reference.title.length > 1024
  ) {
    throw githubError("invalid");
  }
  const now = service.now();
  const stored = service.store.putReference(
    {
      referenceId: referenceId(
        caller.workspaceId,
        ref,
        reference.kind,
        reference.number,
        reference.targetKind,
        id,
      ),
      workspaceId: caller.workspaceId,
      repository: key(ref),
      kind,
      number: Number(reference.number),
      targetKind: target,
      targetId: id,
      title: reference.title,
      revision: 0,
      createdAtMs: now,
      updatedAtMs: now,
    },
    Number(request.expectedRevision),
  );
  return referenceMessage(stored);
}

/**
 * 只删掉一条连接。它指着的远端对象和本地会话都原封不动。
 */
export function unlinkReference(
  service: GithubService,
  caller: Caller,
  request: UnlinkGithubReferenceRequest,
): UnlinkGithubReferenceResponse {
  service.authorize(caller, SCOPE_WRITE);
  const id = request.referenceId;
  if (id === "" || id.length > 256 || request.expectedRevision === 0n) {
    throw githubError("invalid");
  }
  service.store.deleteReference(
    caller.workspaceId,
    id,
    Number(request.expectedRevision),
  );
  return create(UnlinkGithubReferenceResponseSchema, {
    referenceId: id,
    unlinked: true,
  });
}

/**
 * 列这个工作空间的连接，可以只看一个目标。空目标列整个工作空间；它从来不是一次
 * 隐式匹配。
 */
export function listReferences(
  service: GithubService,
  caller: Caller,
  request: ListGithubReferencesRequest,
): ListGithubReferencesResponse {
  service.authorize(caller, SCOPE_READ);
  const limit = pageLimit(Number(request.limit));
  const records = service.store.references(
    caller.workspaceId,
    request.targetId,
    request.afterId,
    limit + 1,
  );
  const result = create(ListGithubReferencesResponseSchema, {});
  for (const [index, record] of records.entries()) {
    if (index === limit) {
      result.hasMore = true;
      break;
    }
    result.references.push(referenceMessage(record));
    result.nextId = record.referenceId;
  }
  return result;
}

/* -------------------------------- 凭据包装 -------------------------------- */
//
// 它们返回的状态**永远不含令牌**；只有一次 configure 请求带着令牌，而且是入站的。

export async function getCredential(
  service: GithubService,
  caller: Caller,
): Promise<GithubCredentialStatus> {
  service.authorize(caller, SCOPE_READ);
  return service.credentials.status();
}

/**
 * `ConfigureCredential` 同时要求 `github:write` **和** settings 权限：选择这台
 * 机器的 GitHub 令牌从哪来是一次设置变更，不是一次普通的 Issue 编辑。
 */
export async function configureCredential(
  service: GithubService,
  caller: Caller,
  request: ConfigureGithubCredentialRequest,
): Promise<GithubCredentialStatus> {
  service.authorize(caller, SCOPE_WRITE);
  service.authorize(caller, "settings:write");
  return service.credentials.configure(
    request.source,
    request.token,
    request.apiBase,
    Number(request.expectedRevision),
  );
}

export async function revokeCredential(
  service: GithubService,
  caller: Caller,
  request: RevokeGithubCredentialRequest,
): Promise<GithubCredentialStatus> {
  service.authorize(caller, SCOPE_WRITE);
  service.authorize(caller, "settings:write");
  return service.credentials.revoke(Number(request.expectedRevision));
}
